### PR TITLE
table: Use `MutateBuffers` to manage buffers

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: ref #54392, ref #54397

### What changed and how does it work?

This PR introduces a better way to manage buffers  in `AddRecord/UpdateRecord/DeleteRecord`. It renamed `TablesBuffer` to `MutateBuffers` and returns buffers by methods instead of accessing the inner object directly. `MutateBuffers` is defined as:

```go
type MutateBuffers struct {
	addRecord    *AddRecordBuffer
	update       *UpdateRecordBuffer
	colSizeDelta *ColSizeDeltaBuffer
}

func NewMutateBuffers(stmtBufs *variable.WriteStmtBufs) *MutateBuffers {
	return &MutateBuffers{
		addRecord: &AddRecordBuffer{
			encodeRowBuffer: encodeRowBuffer{
				writeStmtBufs: stmtBufs,
			},
		},
		updateRecord: &UpdateRecordBuffer{
			encodeRowBuffer: encodeRowBuffer{
				writeStmtBufs: stmtBufs,
			},
		},
		colSizeDelta: &ColSizeDeltaBuffer{},
	}
}

func (b *MutateBuffers) GetAddRecordBufferWithCap(cap int) *AddRecordBuffer {
	buffer := b.addRecord
	buffer.Reset(cap)
	return buffer
}

...
```

The internal buffers are defined as private fields that can only be fetched by methods like `GetAddRecordBufferWithCap` which also does some works to make sure the buffers are inited when using it.  `AddRecordBuffer` also provides some methods:

```go
type AddRecordBuffer struct {
	encodeRowBuffer
}

type encodeRowBuffer struct {
	// colIDs is the column ids for a row to be encoded.
	colIDs []int64
	// row is the column data for a row to be encoded.
	row []types.Datum
	// writeStmtBufs refs the `WriteStmtBufs` in session
	writeStmtBufs *variable.WriteStmtBufs
}


func (b *encodeRowBuffer) AddColVal(colID int64, val types.Datum) {
	b.colIDs = append(b.colIDs, colID)
	b.row = append(b.row, val)
}

func (b *encodeRowBuffer) WriteMemBufferEncoded(
	cfg RowEncodingConfig, loc *time.Location, ec errctx.Context,
	memBuffer kv.MemBuffer, key kv.Key, flags ...kv.FlagsOp,
) error {
	// ... encode the data and write the result to membuffer.
}
```

You can see that the outside code cannot access the inner slices directly. We can call `AddColVal` when we want to add new column data to the buffer and call `WriteMemBufferEncoded` when we want to write a encoded row to membuffer. Because the inner slices are private, we can ensure that those slices will not be cached somewhere else by mistake only if the `WriteMemBufferEncoded` is implemented correctly.

TODO:

Some methods are still exposing the inner slices to the outside such as `GetColDataBuffer` or `GetRowToCheck`, we should avoid it in the future PRs.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by ekexium — APPROVED** [review]
LGTM. It protects the usage of these buffer. I just have some concerns.

The refactoring does limit the access of these buffers to certain methods. But it introduced obvious complexity and reduced flexibility. I'm not sure whether this is the best approach.

There are more allocations in the hot path that can be optimized. I won't expect we do such refactoring work for every buffer that may be introduced.

File: pkg/table/context/buffers.go [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+	return memBuffer.SetWithFlags(key, encoded, flags...)
+}
+
+// AddRecordBuffer is the buffer for AddRecord operation.
+type AddRecordBuffer struct {
+	encodeRowBuffer
+}
+
+// GetColDataBuffer returns the buffer for column data.
+// TODO: make sure the inner buffer is not used outside directly.
```
[the remark this answers is not in the document]
Comment (lcwangchao) [reply pkg/table/context/buffers.go:88]: For example, we can introduce a method `UpdateColSizeDeltaMap(delta *variable.TableDelta)` to update the `TableDelta.ColSize` in the argument without exposing the inner slice directly.

File: pkg/table/context/buffers.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+// GetRowToCheck gets the row data for constraint check.
+// TODO: make sure the inner buffer is not used outside directly.
+func (b *UpdateRecordBuffer) GetRowToCheck() []types.Datum {
+	return b.rowToCheck
+}
+
+// AddColValToCheck adds a column value to the buffer for checking.
+func (b *UpdateRecordBuffer) AddColValToCheck(val types.Datum) {
+	b.rowToCheck = append(b.rowToCheck, val)
+}
```
[the remark this answers is not in the document]
Comment (lcwangchao) [reply pkg/table/context/buffers.go:113]: `rowToCheck` stands for the whole row with type `[]types.Datum`, but this method means adding one column delta to this row.

But I think we can rename the method to `AddColValForRowCheck`.

[the remark this answers is not in the document]
Comment (lcwangchao) [reply pkg/table/context/buffers.go]: I removed `UpdateRecordBuffer` and used a `CheckRowBuffer` for constraint checking. So `CheckRowBuffer.AddColVal` is easy to understand.

**Comment by lcwangchao** [comment]
I've renamed some buffers in the new commit:

```
// MutateBuffers is used to get the buffers for table mutating.
type MutateBuffers struct {
	encodeRow    *EncodeRowBuffer
	checkRow     *CheckRowBuffer
	colSizeDelta *ColSizeDeltaBuffer
}
```

Seems their semantics are more clear and easier to use.

File: pkg/table/context/buffers.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+// GetColSizeDelta gets the column size delta.
+// TODO: make sure the inner buffer is not used outside directly.
+func (b *ColSizeDeltaBuffer) GetColSizeDelta() []variable.ColSize {
+	return b.delta
+}
+
+// MutateBuffers is used to get the buffers for table mutating.
+type MutateBuffers struct {
```
Comment (cfzjywxk) [inline pkg/table/context/buffers.go:132]: It's better to explain in the comments the **lifetime** and expected usage pattern of these buffers as there are so many `buffer structure`, and there is also optimization like memory re-use like https://github.com/pingcap/tidb/pull/53832.

**Review by cfzjywxk — APPROVED** [review]
LGTM

**Pull request merged**